File: taejoong/leetcode/1514_Path_with_Maximum_Probability.java
class Solution {

    static class Node {
        int to;
        double prob;

        Node(int to, double prob) {
            this.to = to;
            this.prob = prob;
        }
    }

    public double maxProbability(int n, int[][] edges, double[] succProb, int start_node, int end_node) {
        List<Node>[] graph = new ArrayList[n];

        for(int i = 0; i < n; i++) {
            graph[i] = new ArrayList<>();
        }

        for(int i = 0; i < edges.length; i++) {
            int a =…           
            if(cur.prob < best[cur.to]) {
                continue;
            }

            if(cur.to == end_node) {
                return cur.prob;
            }

            for(Node next : graph[cur.to]) {
                double newProb = cur.prob * next.prob;

                if(newProb > best[next.to]) {
                    best[next.to] = newProb;

                    pq.offer(new Node(next.to, newProb));
                }
            }
        }
        return 0.0;
    }
}

File: hosung/leetcode/1514. Path with Maximum Probability.py
import heapq

class Solution(object):
    def maxProbability(self, n, edges, succProb, start_node, end_node):

        graph = [[] for _ in range(n)]

        for i in range(len(edges)):
            u, v = edges[i]
            graph[u].append((v, succProb[i]))
            graph[v].append((u, succProb[i]))

        probs = [0] * n
        probs[start_node] = 1

        pq = [(-1, start_node)]

        while pq:

            cur_prob, node = heapq.heappop(pq)
            cur_prob = -cur_prob

            if node == end_node:
                return cur_prob

            if cur_prob < probs[node]:
                continue

            for nxt, p in graph[node]:

                new_prob = cur_prob * p

                if new_prob > probs[nxt]:
                    probs[nxt] = new_prob
                    heapq.heappush(pq, (-new_prob, nxt))

        return 0
